pkg/controller/volume/attachdetach/util/util.go

```diff
@@ -155,9 +156,30 @@ func DetermineVolumeAction(pod *v1.Pod, desiredStateOfWorld cache.DesiredStateOf
 		// should be detached or not
 		return keepTerminatedPodVolume
 	}
+	if isPodNeededToBeRemoved(pod) {
+		// if kubelet is down, we let kubelet policy dictate if volume
+		// should be detached or not
+		return keepTerminatedPodVolume
+	}
+
 	return defaultAction
 }
 
+// deleted pod that should be gone and volumes in it should be removed from dswp if they are not deleted after DeletionGracePeriod seconds
+// this can happen if a node is lost, and the kubelet is never able to confirm deletion.
+func isPodNeededToBeRemoved(pod *v1.Pod) bool {
+	if pod.DeletionTimestamp != nil && pod.DeletionGracePeriodSeconds != nil {
```

Comment:
Do we care about:
```GO
podStatus.Phase == v1.PodFailed || podStatus.Phase == v1.PodSucceeded
```

If pod errors out or completes but hasn't been deleted yet? `IsPodTerminated(...)` currently handles that as a termination state.

Reply:
Pod/Volumes in termination state should be detached. 
If the assumption above is right, it is ok here because both `IsPodTerminated(...)` and `isPodNeededToBeRemoved` are called in `DetermineVolumeAction`.
Correct me please if i am wrong, @saad-ali

Reply:
@saad-ali yeah `IsPodTerminated` handles those as terminated state.

Reply:
Ah, true, in those cases the third statement isn't checked.